charms/reactive/flags.py

```diff
@@ -269,3 +274,86 @@ def get_state(flag, default=None):
        For internal use only.
     """
     return _get_flag_value(flag, default)
+
+
+# INTERNAL
+
+@hookenv.atstart
+def _manage_automatic_flags():
+    _manage_upgrade_flags()
+    _manage_config_flags()
+    _manage_leadership_flags()
+    # TODO: storage flags?
+
+
+def _manage_upgrade_flags():
+    # import here to prevent circular import :(
+    from charms.reactive.helpers import data_changed
+    hook_name = hookenv.hook_name()
+    charm_dir = Path(hookenv.charm_dir())
+
+    if hook_name == 'upgrade-charm':
+        charm_rev = (charm_dir / '.juju-charm').read_text()
+        if data_changed('reactive.charm-rev', charm_rev):
+            clear_flag('upgrade.charm.completed')
+            set_flag('upgrade.charm.completed')
+        clear_flag('upgrade.resources.check')
+        set_flag('upgrade.resources.check')
+
+    if hook_name == 'pre-series-upgrade':
+        clear_flag('upgrade.series.started')
+        set_flag('upgrade.series.started')
+
+    if hook_name == 'post-series-upgrade':
+        clear_flag('upgrade.series.completed')
+        set_flag('upgrade.series.completed')
```

Comment:
> My instinct would be to just use `hook.hookname` as the flag, rather than require charmers to follow your mapping of Juju hook to reactive flag.  ... it just places more burden on charm developers to deal with the split brain.

In short: I think the "split-brain" issue will exist as long as we require charmers to think about both charms.reactive and the hook lifecycle while writing handlers.

I responded more thoroughly on your other comment: https://github.com/juju-solutions/charms.reactive/pull/182#discussion_r197151866

Note that line 308 and 309 have the same flag, this is simply so any triggers get rerun; the framework doesn't remove the `upgrade.series.started` flag automatically.

>  just places more burden on charms.reactive to document the differences between the Juju model

I'm happy to volunteer to do that ;) Although I think Cory already did a good job in [`managed-flags.rst`](https://github.com/juju-solutions/charms.reactive/blob/94399843ab9a81dc94b54c974cf13382337fde7c/docs/managed-flags.rst)

> Similarly upgrade-charm above, which is adding heuristics for the type of upgrade that the charm developer wants to catch, and really just a guess. We can certainly add extra flags to save devs checking the underlying charm revision themselves, but I personally would want to stick with running my upgrade process whenever Juju tells me it might be needed. Eg. after a jujud upgrade, which pulls in the new model config specifying snap proxies, which tells the snap layer to pay attention to the environment change and remove its snapd hacks now Juju has taken control.

I agree that the "normal" `upgrade-charm` flag (regardless of what its called) should be set _at least_ everytime when the the `upgrade-charm` hook runs, since that hook has a lot more "meaning" to it than simply "the charm code was upgraded".